### ISSUE TITLE
issue: DeepMap not working correctly with Union Types

### ISSUE THREAD

**Issue by tatianainama** [react-hook-form/react-hook-form#8815]
### Version Number

7.4.1

### Codesandbox/Expo snack

https://codesandbox.io/s/laughing-pascal-ev9v9x?file=/src/App.tsx

### Steps to reproduce

In this example I'm using `zod` for schema validation.

1. See given Codesandbox
2. Check `useEffect` and `showErrors` for type errors

----

Given the following schema:

```ts
const Gelato = z.literal("gelato");
const Sorbetto = z.literal("sorbetto");

const GelatoSchema = z.object({
  type: Gelato,
  creamFlavor: z.string().min(1)
});

const SorbettoSchema = z.object({
  type: Sorbetto,
  fruitFlavor: z.string().min(1)
});

const FormSchema = z.object({
  name: z.string().min(1),
  favoriteIceCream: z.union([GelatoSchema, SorbettoSchema])
});

type FormSchemaType = z.infer<typeof FormSchema>;
```

and using `useForm` like the following:

```ts
const {
    register,
    handleSubmit,
    control,
    formState: { touchedFields, errors }
  } = useForm<FormSchemaType>({
    defaultValues: {
      name: "",
      favoriteIceCream: {
        type: Gelato.value,
        creamFlavor: ""
      }
    },
    resolver: zodResolver(FormSchema)
  });

  const formState = useWatch({ control });
```

will throw a type error message when trying to access `creamFlavor` or `fruitFlavor` from `touchedFields` or `errors`:

```ts
Property 'creamFlavor' does not exist on type 'DeepMap<{ type: "gelato"; creamFlavor: string; } | { type: "sorbetto"; fruitFlavor: string; }, true>'.
```

### Expected behaviour

I was expecting both `touchedFields` and `errors` have the same type as `formValues`, but only the common field between the two types (in this case, the `type: 'gelato' | 'sorbetto'`) is available 

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
It's a design limitation with union type.

**Comment by tatianainama** [comment]
Hi @bluebill1049, I am confused on how it's a design limitation only for `touchedFields` and `errors` types.
I see there is a different implementation here:
 - `touchedFields` and `errors` are represented as an intersection `&` (that's why only the common fields are available)
 - `formValues` is represented as a union `|` (so all fields are available)

I saw this https://github.com/react-hook-form/react-hook-form/pull/8605 but it was closed. 
Is it ok if I try implement it and open a new one?

**Comment by bluebill1049** [comment]
> Is it ok if I try implement it and open a new one?

if you can find a solution, yes please.

**Comment by felixschorer** [comment]
This usecase shouldn't be supported in my opinion. It would be a foot gun.

If implemented, `getFieldValues` could and will return values like `{ type: 'gelato', fruitFlavor: 'strawberry' }` or  `{ type: 'gelato', fruitFlavor: 'strawberry', creamFlavor: 'chocolate' }`.

This is due to the internal state representation of RHF. It cannot support union types at a from group level.

**Comment by tatianainama** [comment]
@felixschorer thanks for this explanation!

I think this is a valid use case, complex form states like this are currently being handle by RHF correctly, but types are not reflecting what the library does in reality.

RHF distributes values over union types, but types definition are non-distributive, so the types for `touchedFields`, `dirtyFields` and `errors` are **invalid** 

as you mentioned, the data returned from this checkbox group is the following:

```js
// formState.favoriteIceCream.type sorbetto 
// touchedFields:
{
  "favoriteIceCream": {
    "type": true,
    "creamFlavor": true,
    "fruitFlavor": true
  }
}
```

Which I'm currently manually casting to the distributive union to be able to compile the code.

What else do you think it could be done to fix this discrepancy?

**Comment by felixschorer** [comment]
@tatianainama the issue is that there are cases where RHF cannot handle unions at form group level correctly.

Consider a case like this where the `kind` and `value` properties are controlled by individual input fields:
```ts
interface VariantA {
  kind: 'A'
  value: string
}

interface VariantB {
  kind: 'B'
  value: boolean
}

type FieldValues= VariantA | VariantB
```
- In case `kind` is `'A'` a text input would be rendered to change the `value`.
- In case `kind` is `'B'` a checkbox would be rendered to change the `value`.

Switching from one kind to another would result in the text input receiving a boolean value and the checkbox receiving a string value.

---

Another issue would be default values. Using your example from above the default values would have to have this type:
```ts
interface DefaultValues {
  type: 'gelato' | 'sorbetto'
  creamFlavor: string
  fruitFlavor: string
}
```
The default value has to satisfy both variants, otherwise the `fruitFlavor` input would be initialized with an `undefined` value when you change the `type` from `'gelato'` to `'sorbetto'`.

For my above example the type for the default value would need to be:
```ts
interface DefaultValues {
  kind: 'A' | 'B'
  value: string & boolean
}
```
- `string & boolean` evaluates to `never` as there is no overlap between the types.
- How do you decide at a type level when to intersect the types of a property (e.g. `value`) and when to construct a union (e.g. `kind`)?
- How do you handle complex fields, i.e., fields whose values are objects? There's the `NestedValue` helper type to mark those fields as complex fields, but that's something which I'd like to get away from.

---

To sum up, the internal state of RHF in your example actually has this implicit type:
```ts
type FieldValues = {
  name: string
  favoriteIceCream: {
    type: 'gelato' | 'sorbetto'
    creamFlavor: string
    fruitFlavor?: string | undefined
  }
}
```
So, I would recommend that you change your schema to align with that internal state representation:
```ts
const Gelato = z.literal("gelato")
const Sorbetto = z.literal("sorbetto")

const IceCreamSchema = z.object({
  type: z.union([Gelato, Sorbetto]),
  creamFlavor: z.string().min(1),
  fruitFlavor: z.string().min(1)
})

const FormSchema = z.object({
  name: z.string().min(1),
  favoriteIceCream: IceCreamSchema 
});

type FormSchemaType = z.infer<typeof FormSchema>;
```

**Comment by bluebill1049** [comment]
@tatianainama thanks for the feedback, which section do you think would be suitable for such content?
